File: 01-09/adts/table/table.cpp
#pragma once

#include "adts/iterable/iterable.cpp"

template <class K, class V>
class Table : public Iterable<Pair<K, V>>
{
public:
  virtual void add(K key, V value) = 0;
  virtual void remove(K key) = 0;
  virtual int size() = 0;
  virtual bool has(K key) = 0;
  virtual V get(K key) = 0;
};

File: 01-09/adts/table/Table.java
package adts.table;

import adts.pair.Pair;

public interface Table<K, V> extends Iterable<Pair<K, V>> {
  void add(K key, V value);

  void delete(K key);

  int size();

  boolean has(K key);

  V get(K key);

  Iterable<K> keys();

  Iterable<V> values();

}
